File: testQmatrixDomination_CO.py
# -*- coding: utf-8 -*-
"""
Created on Fri Jun 17 16:14:52 2022

@author: VG244206
"""
import numpy as np
import nlopt

def testQmatrixDomination_CO(Q, lVOP):

    # Test if Q <= Qvop (combined domination)
    # input :
    #    Q10g : NcxNc Hermitian matric 
    #    Qvop : list of Nvop NcxNc Hermitian matrices
    # Output :
    #    R       N real>0      max ((V^H Q V)/(V^H Qvop V))
    #    V       NcxN complex    argmax ((V^H Q V)/(V^H Qvop V))

    
    Nvop = len(lVOP);            
    Nc = Q.shape[0];
    
    for Qvop in lVOP:
        assert Qvop.ndim == 2 and Qvop.shape[0]==Nc and Qvop.shape[1]==Nc, 'Bad list of VOP (shape)'
    
    
    # Convert to real symetric matrices
    CHtoRS = lambda Q: np.vstack((np.hstack((np.real(Q), -np.imag(Q))), np.hstack((np.imag(Q), np.real(Q)))));
    
    QvopR = [ CHtoRS(Qvop) for Qvop in lVOP];
    QR = CHtoRS(Q);
    
    SAR = lambda Qvop,V: V.transpose() @ Qvop @ V;
        
    if  Nvop == 1:
            
        # in this case do an eignevalue analysis of QR-QvopR
        # and look for the maximum eigenvalue
        # If that eigenvalue is negative then QR<=QvopR and R[i] will be < 1
         
        H = QR-QvopR[0];
        eigW,eigV = np.linalg.eigh(H);
        VR = eigV[:,2*Nc-1]; # vector corresponding to the max eigenvalue
        
    else:
            
        # In this case run a convex optimization
        
        V0R = np.zeros((2*Nc));
        V0R[0] = 1;
        eigW,eigV = np.linalg.eigh(QR);
        V0R = eigV[:,2*Nc-1];
        C = max([SAR(Qvop,V0R) for Qvop in lVOP]);
        V0R *= 0.5 / (np.sqrt(max(0,C))+1e-9);
        
        # prepare nlopt     
        opt = nlopt.opt(nlopt.LD_MMA , 2*Nc)
        opt.set_min_objective(lambda x,grad: objfun(QR, x, grad))
        opt.add_inequality_mconstraint(lambda res,x,grad: constrfun(QvopR, res, x, grad), 1e-8*np.ones(Nvop))
        opt.set_xtol_rel(1e-6)
        opt.set_maxeval(200)
        
        # find minimizer
        VR = opt.optimize(V0R)
        
    # Store solution of the problem  max ((V^H Q V)/(V^H Qvop V))
    R = SAR(QR,VR) / SAR(QvopR,VR) - 1;
    V = VR[0:Nc] + 1j * VR[Nc:2*Nc];

    return (R,V)

# Objective function

def objfun(Q, V, G) :
    P = -V.transpose() @ Q
    C = P @ V;
    G[:] = 2 * P  #.flatten();
    
    return C;


# Constraints

def constrfun(lQ,C,V,G) :
    
    Vt = V.transpose();
    
    for i,Q in zip(range(0,len(lQ)),lQ) :
        
        P = Vt @ lQ[i];
        C[i] = P @ V - 1;
        G[i,:] = 2*P;

File: testQmatrixDomination_CO.m
function [R, V] = testQmatrixDomination_CO(Q, Qvop)

% Test whether Q <= Qvop (combined domination)
% Input : 
%   Q       NcxNcxN complex Hermitian>0
%   Qvop    NcxNcxNvop complex Hermitian>0
% Output :
%   R       1xN real>0      max ((V^H Q V)/(V^H Qvop V))
%   V       NcxN complex    argmax ((V^H Q V)/(V^H Qvop V))
% Requirement :
%   MATLAB R2020b 
%   Optimization toolbox

N = size(Q, 3);
Nc = size(Qvop, 1);

V = zeros(Nc, N);
R = zeros(1, N);

CHtoRS = @(Q) cat(1, cat(2, real(Q), -imag(Q)), cat(2, imag(Q), real(Q)));
QvopR = CHtoRS(Qvop);

evalSARReal = @(Q,V) max(max(reshape(pagemtimes(V, 'transpose', pagemtimes(Q, V), 'none'), size(Q, 3), 1)),0);

parfor i = 1:N
    
    QtR = CHtoRS(Q(:,:,i));
    QtR = 0.5*(QtR+QtR');
    
    if (size(Qvop, 3)==1)
        
        [eigV, eigD] = eig(QtR-QvopR, 'vector');
        eigD = real(eigD);
        [~, k] = max(eigD);
        VR = real(eigV(:,k));
        
    else
        
        [eigV, eigD] = eig(QtR, 'vector');
        eigD = real(eigD);
        [~, k] = max(eigD);
        V0R = real(eigV(:,k));
        C = max(evalSARReal(QvopR, V0R));
        V0R = 0.5*V0R / (sqrt(C)+eps);
        
        opt = optimset('Algorithm', 'sqp', 'MaxIter', 1000,  'Display', 'off', 'GradObj', 'on', 'GradConstr', 'on'); %,'OutputFcn',outfun);
        
        VR = fmincon(@(VR) objfun(QtR,VR), V0R, [], [], [], [], [], [], @(VR) VopConstr(QvopR, VR), opt);
    end
    
    V(:,i) = VR(1:Nc) + 1i * VR(Nc+1:2*Nc);
    R(i) = evalSARReal(QtR,VR)./evalSARReal(QvopR,VR);
    
    if (mod(i,1000)== 0)
        fprintf('parfor loop : %d / %d; %f\n', i, N, R(i));
    end
    
    
end

function [C, G] = objfun(Q, V)
C = -V' * Q * V;
G = -2 * Q * V;


function [C,Ceq,G,Geq] = VopConstr(Q,V)

Ceq = [];
C = reshape(pagemtimes(V, 'transpose', pagemtimes(Q, V), 'none'), size(Q, 3), 1) - 1;
Geq = [];
G = 2 * reshape(pagemtimes(Q, V), numel(V), size(Q, 3));



